### DISCUSSION TITLE
What's the best practice of satellite in prodcution environment?

### DISCUSSION THREAD

**Discussion by tuhao1020** [apache/skywalking#8889] [Q&A] [answered]
Is satellite ready for production? 
What's the best practice?
Skywalking kubernetes uses deployment to setup satellite instead of satefulset, does it mean that there is a possibility of data loss in the memory queue?

**Comment by wu-sheng** [comment, marked as answer]
It is ready for product. And officially, mmap(file based and stateful) is just one option, it doesn't activate on default.
You could change if you like.
Satellite is not just for data doesn't lose, this is only monitoring data, not data transaction.

**Comment by tuhao1020** [comment]
@wu-sheng Thanks for your reply! My understanding is that there will be tracing data in the memory queue, and if it is a deployment, it may be lost while scaling, so what is the official recommended way to use it?

**Reply by wu-sheng** [reply]
MMAP is not in memory. This is a common tech on Linux. You could learn on your own.
There is always something in memory, and this is not a transaction Database, so there is always something could be lost. 
You shoul have a correct understanding, don't use data lose term randomly. This actually doesn't make sense.
Some one said, Kafka wouldn't lose data if it insteads of gRPC, but really? In tech perspective, it is false too.
The key features of Satellite is scaling, and provides option of mmap to keep more data in disk rather than memory. But it will cost more resources, aka more moneys.

**Reply by tuhao1020** [reply]
You mean that in a production environment, tracing/metrics/logs just monitoring data, even if the data is partially lost, it does not matter. It's usually enough to use default memory-queue if there are no special requirements, right?

**Reply by wu-sheng** [reply]
No, I am not saying that. I mean different teams have different choices, mmap(file queue) is provided for real product env.
But you keep mentioning data lose, I just want to point out this is a wrong requirement. Keeping most data doesn't equals no data lose.

**Reply by tuhao1020** [reply]
Got it! Thanks a lot! 👍
